### DISCUSSION TITLE
Keycloak distributed scenario

### DISCUSSION THREAD

**Discussion by ricamz** [keycloak/keycloak#29731] [Q&A]
Hi all.

I have a scenario where I want to have 100s of keycloak remote instances  to provide authentication to local clients and a central keycloak instance (with low bandwidth to the remote instances) where all users are created (and permissions created).

Searching in the web i found that there is no keycloak feature that provides this, so I found a code that syncronizes users form central instance to each remote instance.

Is this the best approach? Can you point me to the right direction?

Regards

Ricardo

**Comment by codespearhead** [comment]
What problem are you trying to solve with the suggested approach?

**Comment by ricamz** [comment]
Hi Pedro.Thank you for your question.The problema I am trying to solve is the folly: I have several remote sites with the same set of applications (around 6/7 web applications) and the users are locally authenticated and authorized which means that if a user roams between sites (which happens a few times) I need to create the same user in each site and maintain the password in each site.What I want to achieve is that a user can locally authenticate with the same credentials in several sites and several applications.Also I want for the user to authenticate with biometric information for certain operations , also centrally if possible.Some of  the remote sites have some internet connectivity issues as they are in the rural areas of an African country .Regards RicardoOn 22 May 2024, at 16:49, Pedro Aguiar ***@***.***> wrote:﻿
What problem are you trying to solve with this approach?

**Reply by codespearhead** [reply]
What you need is called Identity Brokering.

To make it clearer, let's imagine there are three separate companies using their own Keycloak instances: Gmail, GSheets, and GPhotos.

Gmail would need to configure GSheets' and GPhotos' Keycloak instances as OIDC Identity Providers in its own Keycloak instance, allowing options like "Sign in with GSheets" and "Sign in with GPhotos" on Gmail's login page. The benefit is that existing GSheets and GPhotos users wouldn’t have to manually create an account in Gmail's Keycloak instance.

However, for the convenience to work both ways, GSheets and GPhotos would also need to configure the other two as OIDC Identity Providers. This becomes impractical as the number of instances increases.

To solve this issue, you can create a central Keycloak instance, let’s call it "Google", that acts as an Identity Broker. This central instance would provide a single "Sign in with Google" button that delegates authentication requests to the other Keycloak instances without storing user data itself. This simplifies the configuration and management since each of the three instances would only need to configure Google's Keycloak instance as an OIDC Identity Provider.

**Reply by ricamz** [reply]
Thank you, codespearhead.

If I understood your idea (sorry if I didn't) is to have a central Keycloak instance and then this instance would redirect to the keycloak instance of the remote site.

Using your terminology, in my reality I have GSheets and GPhotos in site 1 in its users, and GSheets and GPhotos in site 2, each site with its own users. How can the central keycloak instance redirect the login based on the username? Because the users can't be related to the sites, they have to be named like bob.smith@google.com.

Or your idea is to have  Keycloak instances in each site to redirect to a central Keycloak instance where all users are managed. That is my idea, but my question is: what to do when there is no internet connectivity between the central location and remote sites? Can we store token with longer valid times (days?) Or is there a method to syncronize the users database between the central keycloak instance and remote site instances?

**Reply by ricamz** [reply]
Hi codespearhead.

So I have a lab doing what you said and it works. But I have a few questions that are specific to my environment issues.

1) Is there a way to automatically redirect a user to the central keycloak instance without having a button like "Sign in with Google". Like when the local Keycloak instance finds that the user doesn't exist it automatically redirects the user login form to the Central Keycloak login form.

2) Is it a good idea to give tokens without expiration ? The reason is that some sites can be without internet for more than 30 days and it is very problematic to have someone that cannot login during that time. My idea was that, when the user authenticates the first time with the central keycloak, if using the same computer/browser it doesn't need to authenticate again, logging in the application automatically.

Thanks for your help.

Regards
Ricardo
